packages/lesswrong/server/serverSentEvents.ts

```diff
@@ -203,3 +208,78 @@ async function checkForTypingIndicators() {
     }
   }
 }
+
+async function checkForActiveDialoguePartners() {
+  const numOpenConnections = Object.keys(openConnections).length;
+  if (!numOpenConnections) {
+    return;
+  }
+
+  for (let userId of Object.keys(openConnections)) {
+    const userDialogues = await new PostsRepo().getUserDialogues(userId);
```

Comment:
We should be starting by fetching all the possibly-live user sessions from `CkEditorUserSessions` first, which is much cheaper than iterating over every single connected user session and fetching all of those user's dialogues.  Something like `CkEditorUserSessions.find({ endedAt: { $exists: false } }).fetch();`?

After that, you fetch the corresponding `userId` and `coauthorStatuses` from the posts based on the live user sessions' `documentId`s: `Posts.find({ _id: { $in: ckEditorUserSessionDocumentIds }, { projection: { userId: 1, coauthorStatuses: 1 } }).fetch()`.  (The projection modifies the generated SQL query to only return those fields, it's like saying `SELECT "userId", "coauthorStatuses"` instead of `SELECT *`; this can reduce latency noticeably because posts are pretty big.)

Then you get the intersection of all the author ids of those posts and the set of userIds in `openConnections` to see which of the open connections you need to send a new message to (along with whatever other logic is appropriate; I haven't really parsed the rest of what's happening here yet).